sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryStorageStreamSource.java

```diff
@@ -17,161 +17,565 @@
-        Storage.StreamPosition streamPosition,
-        SerializableFunction<SchemaAndRowProto, T> parseFn,
-        BoundedSource<T> source,
-        BigQueryServices client,
-        BigQueryOptions options) {
-      this.readSession = checkNotNull(readSession, "readSession");
-      this.client = checkNotNull(client, "client");
-      this.parseFn = checkNotNull(parseFn, "parseFn");
-      this.source = checkNotNull(source, "source");
-      this.options = options;
-
-      this.request = Storage.ReadRowsRequest.newBuilder().setReadPosition(streamPosition).build();
+    private StreamPosition currentPosition;
+    private T currentRecord;
+
+    BigQueryStorageStreamReader(BigQueryStorageStreamSource<T> source, BigQueryOptions options)
+        throws IOException {
+      this.currentSource = source;
+      this.client = source.bqServices.getStorageClient(options);
+      this.parseFn = source.parseFn;
+      this.schema = source.readSession.getProjectedSchema();
+      this.rangeTracker =
+          new StreamOffsetRangeTracker(
+              source.stream,
+              source.startOffset,
+              source.stopOffset,
+              source.splitDisposition != SplitDisposition.SELF);
     }
 
-    /**
-     * Empty operation, the table is already open for read.
-     *
-     * @throws IOException on failure
-     */
     @Override
     public boolean start() throws IOException {
-      responseIterator = client.getTableReadService(options).readRows(request);
-      return advance();
+      return (startImpl() && rangeTracker.tryReturnRecordAt(true, currentPosition))
+          || rangeTracker.markDone();
+    }
+
+    private boolean startImpl() {
+      BigQueryStorageStreamSource<T> source = getCurrentSource();
+      LOG.info("Starting reader {} from source {}", this, source);
+      if (source.splitDisposition == SplitDisposition.SELF) {
+        currentPosition = rangeTracker.getStartPosition();
+      } else {
+        currentPosition = processSplit(source.splitPosition, source.splitDisposition);
+      }
+
+      ReadRowsRequest request =
+          ReadRowsRequest.newBuilder().setReadPosition(currentPosition).build();
+      responseStream = client.readRowsCallable().call(request);
+      responseIterator = responseStream.iterator();
+      rowIterator = null;
+      return readNextRow();
     }
 
     @Override
-    public boolean advance() {
-      if (rowIterator == null || !rowIterator.hasNext()) {
+    public boolean advance() throws IOException {
+
+      currentPosition =
+          currentPosition.toBuilder().setOffset(currentPosition.getOffset() + 1).build();
+
+      while (true) {
+        if (!readNextRow()) {
+          return rangeTracker.markDone();
+        }
+
+        if (rangeTracker.tryReturnRecordAt(true, currentPosition)) {
+          return true;
+        }
+
+        if (rangeTracker.isDone()) {
+          return false;
+        }
+
+        if (responseStream != null) {
```

Comment:
Nit: Might be good to add a comment above that mentions something like:

```
The only time tryReturnRecordAt returns fails (and the rangeTracker is not done) is when a split is pending.
```

Reply:
I incorporated this feedback into https://github.com/kmjung/beam/pull/10.